packages/tailwindcss/src/index.ts

```diff
@@ -311,30 +297,9 @@ export async function compile(
     customUtility(designSystem)
   }
 
-  let api: PluginAPI = {
-    addVariant(name, variant) {
-      // Single selector
-      if (typeof variant === 'string') {
-        designSystem.variants.static(name, (r) => {
-          r.nodes = [rule(variant, r.nodes)]
-        })
-      }
-
-      // Multiple parallel selectors
-      else if (Array.isArray(variant)) {
-        designSystem.variants.static(name, (r) => {
-          r.nodes = variant.map((selector) => rule(selector, r.nodes))
-        })
-      }
-
-      // CSS-in-JS object
-      else if (typeof variant === 'object') {
-        designSystem.variants.fromAst(name, objectToAst(variant))
-      }
-    },
-  }
+  let pluginApi = buildPluginApi(designSystem)
 
-  await Promise.all(pluginLoaders.map((loader) => loader.then((plugin) => plugin(api))))
+  await Promise.all(pluginLoaders.map((loader) => loader.then((plugin) => plugin(pluginApi))))
```

Comment:
It was like this before this too. Same thing with v3. I think a hard crash is fine. Users are more likely to see it and try fixing it if it's broken. If it's a warning then it could be more easily missed.